### PR TITLE
LUCENE-9578: TermRangeQuery empty string lower bound edge case

### PR DESCRIPTION
**Pull request by cbuescher**
# Description

Currently a TermRangeQuery with the empty String ("") as lower bound and
includeLower=false leads internally constructs an Automaton that doesn't match
anything. This is unexpected expecially for open upper bounds where any string
should be considered to be "higher" than the empty string.

# Solution

This PR changes "Automata#makeBinaryInterval" so that for an empty string lower
bound and an open upper bound, any String should match the query regardless or
the includeLower flag.

# Tests

Added two new tests to `TestAutomaton`.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `./gradlew check`
- [x] I have added tests for my changes.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/util/automaton/Automata.java [hunk 1]
```diff
@@ -254,7 +254,7 @@ public static Automaton makeBinaryInterval(BytesRef min, boolean minInclusive, B
       cmp = min.compareTo(max);
     } else {
       cmp = -1;
-      if (min.length == 0 && minInclusive) {
+      if (min.length == 0) {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/util/automaton/Automata.java:257]: this looks wrong as we should still make sure that the empty string is rejected if min=="" and minInclusive==false?

**Comment by cbuescher** [comment]
@jpountz thanks for the review, I added a commit that rejects the empty string when "includeMin == false" and added tests for that edge case as well.

File: lucene/core/src/java/org/apache/lucene/util/automaton/Automata.java [hunk 2]
```diff
@@ -85,7 +85,15 @@ public static Automaton makeAnyBinary() {
...
     return a;
   }
-  
+
+  /**
+   * Returns a new (deterministic) automaton that accepts all binary terms except
+   * the empty string.
+   */
+  public static Automaton makeAnyBinaryExceptEmpty() {
+    return makeCharRange(0, 255);
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/util/automaton/Automata.java:94]: I don't think this is right since this only accepts strings of length 1.

Comment (cbuescher) [reply lucene/core/src/java/org/apache/lucene/util/automaton/Automata.java]: Yes, I was missing that this doesn't include a self-referencing accept state. Fixed that and added more tests to check this is working now.

**Comment by cbuescher** [comment]
@jpountz thanks for the review, I added another commit that should fix the problem you detected.

File: lucene/core/src/java/org/apache/lucene/util/automaton/Automata.java [hunk 3]
```diff
@@ -85,7 +85,22 @@ public static Automaton makeAnyBinary() {
...
+
+  /**
+   * Returns a new (deterministic) automaton that accepts all binary terms except
+   * the empty string.
+   */
+  public static Automaton makeAnyBinaryExceptEmpty() {
+    Automaton a = new Automaton();
+    int s1 = a.createState();
+    int s2 = a.createState();
+    a.setAccept(s2, true);
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/util/automaton/Automata.java:97]: Do we need to call setAccept on s1 too? Or is it done implicitly elsewhere (I'm not completely familiar with the Automaton code)

File: lucene/core/src/java/org/apache/lucene/util/automaton/Automata.java [hunk 4]
```diff
@@ -85,7 +85,22 @@ public static Automaton makeAnyBinary() {
     a.finishState();
     return a;
   }
-  
+
+  /**
+   * Returns a new (deterministic) automaton that accepts all binary terms except
+   * the empty string.
+   */
+  public static Automaton makeAnyBinaryExceptEmpty() {
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/util/automaton/Automata.java:93]: Maybe call `makeNonEmptyBinary`?

File: lucene/core/src/java/org/apache/lucene/util/automaton/Automata.java (hunk 3 above)
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/util/automaton/Automata.java:97]: Oops ignore this comment I was totally confused. :)

**Comment by jpountz** [comment]
I had a deeper look to check whether this could be an indicator of other bugs lurking, but it doesn't seem so and the fix looks good to me. For reference here's what the automaton looks like today with an empty string as a min bound and includeMin==false before the above fix.
empty

**Comment by cbuescher** [comment]
@jpountz thanks, I renamed the method. btw. would this change also be backported to the latest 8.7 branch so we can use it in ES to fix https://github.com/elastic/elasticsearch/issues/63386?

**Comment by jpountz** [comment]
The 8.7 branch has not been cut yet so this will be in 8.7. This change looked safe to me so I felt free to merge in spite of the branch being imminently cut.
